File: Day07/Part01/day07_01.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.Arrays; 
import java.util.Map;
import java.util.HashMap;

class day07_01 {


    public static void main (String[] args) {

        System.out.println("Starting Day 07 Part 01...");

        HashMap<String, Program> programs = getPrograms("input.txt");

        System.out.println("Got programs...");

        // finding the base program by finding the program with the largest depth
        Integer biggestDepth = -1;
        Program baseProgram = null;

        for (Map.Entry<String, Program> entry : programs.entrySet()) {
            // System.out.println("New Program: " + entry.getKey());
            // entry.getValue().print();
            // System.out.println("Depth: " + getDepth(entry.getValue(), programs));
            Integer currentDepth = getDepth(entry.getValue(), programs);
            if (currentDepth > biggestDepth) {
                biggestDepth = currentDepth;
                baseProgram = entry.getValue();
            }
        };

        System.out.println("Base Program: " + baseProgram.getName());

    }


    public static HashMap<String, Program> getPrograms(String filename) {
        HashMap<String, Program> programs = new HashMap<String, Program>();
        try {
            BufferedReader br = new BufferedReader(new FileReader(filename));
            String next = br.readLine();
            while (next != null) {
                String[] stringProgram = next.split("\\W+");
                Program temp;
                if (stringProgram.length == 1) {
                    temp = new Program(stringProgram[0], null, null);
                } else if (stringProgram.length == 2) {
                    temp = new Program(stringProgram[0], Integer.valueOf(stringProgram[1]), null);
                } else if (stringProgram.length > 2) {
                    temp = new Program(stringProgram[0], Integer.valueOf(stringProgram[1]), Arrays.copyOfRange(stringProgram, 2, stringProgram.length));
                } else {
                    temp = null;
                }
                programs.put(temp.getName(), temp);
                next = br.readLine();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        if (programs.size() == 0) return null;
        return programs;
    }


    public static Integer getDepth(Program p, HashMap<String, Program> programs) {
        String[] stack = p.getStackedPrograms();
        if (stack == null) {
            return 0;
        } else {
            Integer biggestDepth = 0;
            for (String s : stack) {
                Program p2 = programs.get(s);
                Integer depth = getDepth(p2, programs);
                biggestDepth = (depth > biggestDepth) ? depth : biggestDepth;
            }
            return biggestDepth + 1;
        }
    }
    
}


class Program {

    private String mName;
    private Integer mWeight;
    private String[] mStackedPrograms;

    public Program(String name, Integer weight, String[] programs) {
        this.mName = name;
        this.mWeight = weight;
        this.mStackedPrograms = programs;
    }

    public String getName() {
        return mName;
    }

    public String[] getStackedPrograms() {
        return mStackedPrograms;
    }

    public void print() {
        System.out.println("Name: " + mName);
        System.out.println("Weight: " + mWeight);
        if (mStackedPrograms != null) {
            System.out.print("Stack: [ ");
            for (String s : mStackedPrograms) {
                System.out.print(s + " ");
            }
            System.out.print("]\n\n");
        } else {
            System.out.print("\n\n");
        }
    }

}

File: Day07/Part01/day07_01.py
import re


class Program:

    def __init__(self, name, weight, programs):
        self.name = name
        self.weight = weight
        self.programs = programs

    @classmethod
    def fromList(cls, string_list):
        name = string_list[0]
        weight = string_list[1]
        programs = []
        for i in range(2, len(string_list)):
            if (string_list[i] != ''):
                programs.append(string_list[i])
        return cls(name, weight, programs)


def read_input(filename):
    f = open(filename, "r")
    programs = {}
    line = f.readline()
    while (line):
        split_line = re.split("\\W+", line)
        program = Program.fromList(split_line)
        programs[program.name] = program
        line = f.readline()
    f.close()
    return programs


def get_depth(program, program_dict):
    if (program.programs) == []:
        return 0
    else:
        depth = 0
        for s in program.programs:
            d = get_depth(program_dict[s], program_dict)
            if (d > depth):
                depth = d
        return 1 + depth


if __name__ == "__main__":
    
    programs = read_input("input.txt")
    
    # find base as program with greatest depth
    depth = -1
    program = None
    for k,v in programs.items():
        d = get_depth(v, programs)
        if (d > depth):
            depth = d
            program = v
    
    # print result
    print("Answer:", program.name)
